decimal/bcd.go

```diff
@@ -0,0 +1,229 @@
+// "-". Unused bytes and nibbles to the left of the number are set to 0. scale
+// is set to the scale of the number (this is the exponent, negated).
+func EncodeStringToBCD(buf string) ([]byte, error) {
+	signByte := bytePlus // By default number is positive.
+	if buf[0] == '-' {
+		signByte = byteMinus
+	}
+
+	l := GetNumberLength(buf)
+	if l == 0 {
+		return []byte{signByte}, nil
+	}
+
+	// The first nibble should contain 0, if the decimal number has an even
+	// number of digits. Therefore highNibble is false when decimal number
+	// is even.
+	highNibble := true
+	if l%2 == 0 {
+		highNibble = false
+	}
+	scale := 0 // By default decimal number is integer.
+	var byteBuf []byte
+	for i, ch := range []byte(buf) {
+		// Skip leading zeros.
+		if (len(byteBuf) == 0) && ch == '0' {
+			continue
+		}
+		if (i == 0) && (ch == '-' || ch == '+') {
+			continue
+		}
+		if ch == '.' && scale != 0 {
+			return nil, fmt.Errorf("Number contains more than one point")
+		}
+		// Calculate a number of digits after the decimal point.
+		if ch == '.' {
+			scale = len(buf) - i - 1
+			continue
+		}
+
+		if ch < '0' || ch > '9' {
+			return nil, fmt.Errorf("Failed to convert symbol '%c' to a digit", ch)
+		}
+		digit := byte(ch - '0')
+		if highNibble {
+			// Add a digit to a high nibble.
+			digit = digit << 4
+			byteBuf = append(byteBuf, digit)
+			highNibble = false
+		} else {
+			if len(byteBuf) == 0 {
+				byteBuf = make([]byte, 1)
+			}
+			// Add a digit to a low nibble.
+			lowByteIdx := len(byteBuf) - 1
+			byteBuf[lowByteIdx] = byteBuf[lowByteIdx] | digit
+			highNibble = true
+		}
+	}
+	if highNibble {
+		// Put a sign to a high nibble.
+		byteBuf = append(byteBuf, signByte)
+	} else {
+		// Put a sign to a low nibble.
+		lowByteIdx := len(byteBuf) - 1
+		byteBuf[lowByteIdx] = byteBuf[lowByteIdx] | signByte
+	}
+	byteBuf = append([]byte{byte(scale)}, byteBuf...)
+
+	return byteBuf, nil
+}
+
+// DecodeStringFromBCD converts a BCD Packed Decimal to a string buffer.
+//
+// The BCD packed decimal byte array, together with an associated scale, is
+// converted to a string. The BCD array is assumed full of digits, and must be
+// ended by a 4-bit sign nibble in the least significant four bits of the final
+// byte. The scale is used (negated) as the exponent of the decimal number.
+// Note that zeros may have a sign and/or a scale.
+func DecodeStringFromBCD(bcdBuf []byte) (string, error) {
+	const scaleIdx = 0 // Index of a byte with scale.
+	scale := int(bcdBuf[scaleIdx])
+	// Get a BCD buffer without a byte with scale.
+	bcdBuf = bcdBuf[scaleIdx+1:]
+	length := len(bcdBuf)
+	var digits []string
+	for i, bcdByte := range bcdBuf {
+		highNibble := int(bcdByte >> 4)
+		if !(len(digits) == 0 && highNibble == 0) || len(bcdBuf) == 1 {
+			digits = append(digits, strconv.Itoa(highNibble))
+		}
+		lowNibble := int(bcdByte & 0x0f)
+		if !(len(digits) == 0 && lowNibble == 0) && i != length-1 {
```

Comment:
Please add comments to this condition.

Reply:
Simplified condition and added a comment:
```diff
--- a/decimal/bcd.go
+++ b/decimal/bcd.go
@@ -186,13 +186,13 @@ func DecodeStringFromBCD(bcdBuf []byte) (string, error) {
        length := len(bcdBuf)
        var digits []string
        for i, bcdByte := range bcdBuf {
-               highNibble := int(bcdByte >> 4)
-               if !(len(digits) == 0 && highNibble == 0) || len(bcdBuf) == 1 {
-                       digits = append(digits, string('0' + byte(highNibble)))
-               }
-               lowNibble := int(bcdByte & 0x0f)
-               if !(len(digits) == 0 && lowNibble == 0) && i != length-1 {
-                       digits = append(digits, string('0' + byte(lowNibble)))
+               highNibble := bcdByte >> 4
+               digits = append(digits, string('0'+highNibble))
+               lowNibble := bcdByte & 0x0f
+               // lowNibble of last byte is ignored because it contains a
+               // sign.
+               if i != length-1 {
+                       digits = append(digits, string('0'+lowNibble))
                }
        }
 
```